modules/extensions-support/src/main/java/org/mule/runtime/module/extension/internal/loader/enricher/JavaOAuthDeclarationEnricher.java

```diff
@@ -0,0 +1,177 @@
+import org.mule.runtime.api.util.Reference;
+import org.mule.runtime.extension.api.annotation.Expression;
+import org.mule.runtime.extension.api.annotation.connectivity.oauth.OAuthCallbackValue;
+import org.mule.runtime.extension.api.annotation.connectivity.oauth.OAuthParameter;
+import org.mule.runtime.extension.api.connectivity.oauth.OAuthModelProperty;
+import org.mule.runtime.extension.api.connectivity.oauth.OAuthParameterModelProperty;
+import org.mule.runtime.extension.api.declaration.type.ExtensionsTypeLoaderFactory;
+import org.mule.runtime.extension.api.exception.IllegalConnectionProviderModelDefinitionException;
+import org.mule.runtime.extension.api.loader.ExtensionLoadingContext;
+import org.mule.runtime.module.extension.internal.loader.java.property.DeclaringMemberModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.OperationExecutorModelProperty;
+import org.mule.runtime.module.extension.internal.loader.java.property.oauth.OAuthCallbackValuesModelProperty;
+import org.mule.runtime.module.extension.internal.runtime.connectivity.oauth.UnauthorizeOperationExecutor;
+
+import java.lang.reflect.Field;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Enriches OAuth enabled {@link ConnectionProviderDeclaration}s with further
+ * properties and behaviour which are specific to the Java runtime implementation of the
+ * Extensions API
+ *
+ * @since 4.0
+ */
+public class JavaOAuthDeclarationEnricher extends AbstractAnnotatedDeclarationEnricher {
+
+  private final ClassTypeLoader typeLoader = ExtensionsTypeLoaderFactory.getDefault().createTypeLoader();
+  private final MetadataType stringType = typeLoader.load(String.class);
+  private final MetadataType voidType = typeLoader.load(Void.class);
+  private final MetadataType nullAttributesType = typeLoader.load(NullAttributes.class);
+
+  @Override
+  public void enrich(ExtensionLoadingContext extensionLoadingContext) {
+    final ExtensionDeclaration extensionDeclaration = extensionLoadingContext.getExtensionDeclarer().getDeclaration();
+
+    Set<Reference<ConnectionProviderDeclaration>> visitedProviders = new HashSet<>();
+    Set<Reference<ConfigurationDeclaration>> oauthConfigs = new HashSet<>();
+    Reference<Boolean> oauthGloballySupported = new Reference<>(false);
+
+    new DeclarationWalker() {
+
+      @Override
+      protected void onConnectionProvider(ConnectedDeclaration owner, ConnectionProviderDeclaration declaration) {
+        if (!visitedProviders.add(new Reference<>(declaration))) {
+          return;
+        }
+
+        if (declaration.getModelProperty(OAuthModelProperty.class).isPresent()) {
+          if (owner instanceof ExtensionDeclaration) {
+            oauthGloballySupported.set(true);
+            stop();
+          } else if (owner instanceof ConfigurationDeclaration) {
+            oauthConfigs.add(new Reference<>((ConfigurationDeclaration) owner));
```

Comment:
Making sure that the set operates on references and it's not affected by equals and hashCode implementations